Readme.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Memcached module contains the following components: MemcachedLock and MemcachedCache for working with locks and cache on the Memcached server.

The module contains the following packages:
- **Build** - a standard factory for constructing components
- **Cache** - cache Components in Memcached
- **Lock** - components of working with locks in Memcached

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/concepts/configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-memcached-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)


## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-memcached-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**, **Levichev Dmitry** and **Danil Prisiazhnyi**.

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# Memcached components for Pip.Services in Golang Changelog

## <a name="1.0.2"></a> 1.0.2 (2022-07-10) 

- Updated dependencies
- Fixed logger

## <a name="1.0.0"></a> 1.0.0 (2022-07-10) 
       
### Features
* **build** standard factory for constructing components
* **cahce** cache Components in Memcached
* **lock** components of working with locks in Memcached



test/fixture/LockFixture.go
package test_fixture

import (
	"context"
	"testing"

	clock "github.com/pip-services3-gox/pip-services3-components-gox/lock"
	"github.com/stretchr/testify/assert"
)

const (
	LOCK1 string = "lock_1"
	LOCK2 string = "lock_2"
	LOCK3 string = "lock_3"
)

type LockFixture struct {
	lock clock.ILock
}

func NewLockFixture(lock clock.ILock) *LockFixture {
	c := LockFixture{}
	c.lock = lock
	return &c
}

func (c *LockFixture) TestTryAcquireLock(t *testing.T) {

	ctx := context.Background()

	// Try to acquire lock for the first time
	result, err := c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	// Try to acquire lock for the second time
	result, err = c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.False(t, result)

	// Release the lock
	err = c.lock.ReleaseLock(ctx, "", LOCK1)
	assert.Nil(t, err)

	// Try to acquire lock for the third time
	result, err = c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	c.lock.ReleaseLock(ctx, "", LOCK1)
}

func (c *LockFixture) TestAcquireLock(t *testing.T) {

	ctx := context.Background()

	// Acquire lock for the first time
	err := c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.Nil(t, err)

	// Acquire lock for the second time
	err = c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.NotNil(t, err)

	// Release the lock
	err = c.lock.ReleaseLock(ctx, "", LOCK2)
	assert.Nil(t, err)

	// Acquire lock for the third time
	err = c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.Nil(t, err)

	c.lock.ReleaseLock(ctx, "", LOCK2)
}

func (c *LockFixture) TestReleaseLock(t *testing.T) {
	ctx := context.Background()

	// Acquire lock for the first time
	result, err := c.lock.TryAcquireLock(ctx, "", LOCK3, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	// Release the lock for the first time
	err = c.lock.ReleaseLock(ctx, "", LOCK3)
	assert.Nil(t, err)
	// Release the lock for the second time
	err = c.lock.ReleaseLock(ctx, "", LOCK3)
	assert.Nil(t, err)
}


test/fixture/CacheFixture.go
package test_fixture

import (
	"testing"
	"time"

	ccache "github.com/pip-services3-go/pip-services3-components-go/cache"
	"github.com/stretchr/testify/assert"
)

const (
	KEY1   string = "key1"
	KEY2   string = "key2"
	VALUE1 string = "value1"
	VALUE2 string = "value2"
)

type CacheFixture struct {
	cache ccache.ICache
}

func NewCacheFixture(cache ccache.ICache) *CacheFixture {
	c := CacheFixture{}
	c.cache = cache
	return &c
}

func (c *CacheFixture) TestStoreAndRetrieve(t *testing.T) {

	_, err := c.cache.Store("", KEY1, VALUE1, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store("", KEY2, VALUE2, 5000)
	assert.Nil(t, err)

	select {
	case <-time.After(500 * time.Millisecond):
	}

	val, err := c.cache.Retrieve("", KEY1)

	assert.Nil(t, err)
	assert.NotNil(t, val)
	assert.Equal(t, VALUE1, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY2, &str)
	assert.NotNil(t, result)
	assert.Nil(t, err)
	assert.Equal(t, VALUE2, str)
}

func (c *CacheFixture) TestRetrieveExpired(t *testing.T) {

	_, err := c.cache.Store("", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	select {
	case <-time.After(1500 * time.Millisecond):
	}

	val, err := c.cache.Retrieve("", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY1, &str)
	assert.Nil(t, result)
	assert.Nil(t, err)
}

func (c *CacheFixture) TestRemove(t *testing.T) {
	_, err := c.cache.Store("", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	err = c.cache.Remove("", KEY1)
	assert.Nil(t, err)

	val, err := c.cache.Retrieve("", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)

	var str string
	result, err := c.cache.RetrieveAs("", KEY1, &str)
	assert.Nil(t, result)
	assert.Nil(t, err)
}


test/lock/MemcachedLock_test.go
package test_lock

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	memlock "github.com/pip-services3-gox/pip-services3-memcached-gox/lock"
	memfixture "github.com/pip-services3-gox/pip-services3-memcached-gox/test/fixture"
)

func TestMemcachedLock(t *testing.T) {
	var lock *memlock.MemcachedLock
	var fixture *memfixture.LockFixture

	ctx := context.Background()

	host := os.Getenv("MEMCACHED_SERVICE_HOST")
	if host == "" {
		host = "localhost"
	}

	port := os.Getenv("MEMCACHED_SERVICE_PORT")
	if port == "" {
		port = "11211"
	}

	lock = memlock.NewMemcachedLock()

	config := cconf.NewConfigParamsFromTuples(
		"connection.host", host,
		"connection.port", port,
	)
	lock.Configure(ctx, config)
	fixture = memfixture.NewLockFixture(lock)

	lock.Open(ctx, "")
	defer lock.Close(ctx, "")

	t.Run("Try Acquire Lock", fixture.TestTryAcquireLock)
	t.Run("Acquire Lock", fixture.TestAcquireLock)
	t.Run("Release Lock", fixture.TestReleaseLock)
}


cache/MemcachedCache.go
package cache

import (
	"context"
	"strconv"
	"time"

	"github.com/bradfitz/gomemcache/memcache"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cconv "github.com/pip-services3-gox/pip-services3-commons-gox/convert"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	ccon "github.com/pip-services3-gox/pip-services3-components-gox/connect"
	clog "github.com/pip-services3-gox/pip-services3-components-gox/log"
)

/*
MemcachedCache are distributed cache that stores values in Memcaches caching service.

The current implementation does not support authentication.

Configuration parameters:

 - connection(s):
   - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
   - host:                  host name or IP address
   - port:                  port number
   - uri:                   resource URI or connection string with all parameters in it
 - options:
   - max_size:              maximum number of values stored in this cache (default: 1000)
   - max_key_size:          maximum key length (default: 250)
   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
   - max_value:             maximum value length (default: 1048576)
   - pool_size:             pool size (default: 5)
   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
   - retries:               number of retries (default: 3)
   - timeout:               default caching timeout in milliseconds (default: 1 minute)
     - failures:              number of failures before stop retrying (default: 5)
     - retry:                 retry timeout in milliseconds (default: 30 sec)
     - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)

References:

- *:discovery:*:*:1.0    (optional) IDiscovery services to resolve connection

Example:
	ctx := context.Background()

    cache := NewMemcachedCache[string]();
    cache.Configure(ctx, cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 11211,
    ));

    err := cache.Open(ctx, "123")
      ...


    ret, err := cache.Store(ctx, "123", "key1", []byte("ABC"))
    if err != nil {
    	...
    }

    res, err := cache.Retrive(ctx, "123", "key1")
    value, _ := res.([]byte)
    fmt.Println(string(value))     // Result: "ABC"

*/
type MemcachedCache[T any] struct {
	connectionResolver *ccon.ConnectionResolver
	// maxKeySize         int
	// maxExpiration      int64
	// maxValue           int64
	// poolSize           int
	// reconnect          int
	timeout int
	// retries            int
	// failures           int
	// retry              int
	remove bool
	//idle   int
	client    *memcache.Client
	convertor cconv.IJSONEngine[T]
	logger    clog.CompositeLogger
}

// NewMemcachedCache method are creates a new instance of this cache.
func NewMemcachedCache[T any]() *MemcachedCache[T] {
	c := &MemcachedCache[T]{
		connectionResolver: ccon.NewEmptyConnectionResolver(),
		// maxKeySize:         250,
		// maxExpiration:      2592000,
		// maxValue:           1048576,
		// poolSize:           5,
		// reconnect:          10000,
		timeout: 5000,
		// retries:            5,
		// failures:           5,
		// retry:              30000,
		remove: false,
		//idle:   5000,
		client:    nil,
		convertor: cconv.NewDefaultCustomTypeJsonConvertor[T](),
		logger:    *clog.NewCompositeLogger(),
	}
	return c
}

// Configure method are configures component by passing configuration parameters.
// 	 - ctx context.Context
//   - config    configuration parameters to be set.
func (c *MemcachedCache[T]) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.connectionResolver.Configure(ctx, config)
	c.logger.Configure(ctx, config)

	// c.maxKeySize = config.GetAsIntegerWithDefault("options.max_key_size", c.maxKeySize)
	// c.maxExpiration = config.GetAsLongWithDefault("options.max_expiration", c.maxExpiration)
	// c.maxValue = config.GetAsLongWithDefault("options.max_value", c.maxValue)
	// c.poolSize = config.GetAsIntegerWithDefault("options.pool_size", c.poolSize)
	// c.reconnect = config.GetAsIntegerWithDefault("options.reconnect", c.reconnect)
	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	// c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	// c.failures = config.GetAsIntegerWithDefault("options.failures", c.failures)
	// c.retry = config.GetAsIntegerWithDefault("options.retry", c.retry)
	// c.remove = config.GetAsBooleanWithDefault("options.remove", c.remove)
	//c.idle = config.GetAsIntegerWithDefault("options.idle", c.idle)
}

// SetReferences are sets references to dependent components.
// 	 - ctx context.Context
//   - references 	references to locate the component dependencies.
func (c *MemcachedCache[T]) SetReferences(ctx context.Context, references cref.IReferences) {
	c.connectionResolver.SetReferences(ctx, references)
	c.logger.SetReferences(ctx, references)
}

// IsOpen Checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *MemcachedCache[T]) IsOpen() bool {
	return c.client != nil
}

// Open method are opens the component.
// Parameters:
// 	 - ctx context.Context
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *MemcachedCache[T]) Open(ctx context.Context, correlationId string) error {
	connections, err := c.connectionResolver.ResolveAll(correlationId)

	if err == nil && len(connections) == 0 {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
	}

	if err != nil {
		return err
	}

	var servers []string = make([]string, 0)
	for _, connection := range connections {
		host := connection.Host()
		port := connection.Port()
		if port == 0 {
			port = 11211
		}

		servers = append(servers, host+":"+strconv.FormatInt(int64(port), 10))
	}

	// options = {
	//     maxKeySize: c.maxKeySize,
	//     maxExpiration: c.maxExpiration,
	//     maxValue: c.maxValue,
	//     poolSize: c.poolSize,
	//     reconnect: c.reconnect,
	//     timeout: c.timeout,
	//     retries: c.retries,
	//     failures: c.failures,
	//     retry: c.retry,
	//     remove: c.remove,
	//     idle: c.idle
	// };

	c.client = memcache.New(servers...)
	c.client.Timeout = time.Duration(c.timeout) * time.Millisecond
	//c.client.MaxIdleConns = c.idle

	return nil
}

// Close method are closes component and frees used resources.
// Parameters:
//   - ctx context.Context
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *MemcachedCache[T]) Close(ctx context.Context, correlationId string) error {
	c.client = nil
	return nil
}

func (c *MemcachedCache[T]) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// Retrieve method are retrieves cached value from the cache using its key.
// If value is missing in the cache or expired it returns nil.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//  Retruns: cached value or error.
func (c *MemcachedCache[T]) Retrieve(ctx context.Context, correlationId string, key string) (value T, err error) {
	var defaultValue T

	if state, err := c.checkOpened(correlationId); !state {
		return defaultValue, err
	}
	item, err := c.client.Get(key)
	if err != nil && err == memcache.ErrCacheMiss {
		err = nil
	}
	if item != nil {
		value, err := c.convertor.FromJson(string(item.Value))
		if err != nil {
			return defaultValue, err
		}
		return value, nil
	}
	return defaultValue, err
}

// Store method are stores value in the cache with expiration time.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//   - value             a value to store.
//   - timeout           expiration timeout in milliseconds.
// Returns: error or nil for success
func (c *MemcachedCache[T]) Store(ctx context.Context, correlationId string, key string, value T, timeout int64) (result T, err error) {
	var defaultValue T

	if state, err := c.checkOpened(correlationId); !state {
		return defaultValue, err
	}

	timeoutInSec := int32(timeout) / 1000

	jsonVal, err := c.convertor.ToJson(value)
	if err != nil {
		return defaultValue, err
	}

	item := memcache.Item{
		Key:        key,
		Value:      []byte(jsonVal),
		Expiration: timeoutInSec,
	}
	return value, c.client.Set(&item)
}

// Remove method are removes a value from the cache by its key.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
// Retruns: an error or nil for success
func (c *MemcachedCache[T]) Remove(ctx context.Context, correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)

	if !state {
		return err
	}

	err = c.client.Delete(key)
	if err != nil && err == memcache.ErrCacheMiss {
		err = nil
	}
	return err
}

// Contains check is value stores
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
func (c *MemcachedCache[T]) Contains(ctx context.Context, correlationId string, key string) bool {
	state, err := c.checkOpened(correlationId)
	if !state {
		c.logger.Error(ctx, correlationId, err, "Connection is not opened")
		return false
	}

	if _, err := c.client.Get(key); err != nil {
		return false
	}

	return true
}


lock/MemcachedLock.go
package lock

import (
	"context"
	"strconv"
	"time"

	"github.com/bradfitz/gomemcache/memcache"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	ccon "github.com/pip-services3-gox/pip-services3-components-gox/connect"
	clock "github.com/pip-services3-gox/pip-services3-components-gox/lock"
)

/*
MemcachedLock are distributed lock that implemented based on Memcaches caching service.

The current implementation does not support authentication.

Configuration parameters:

- connection(s):
  - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
  - host:                  host name or IP address
  - port:                  port number
  - uri:                   resource URI or connection string with all parameters in it
- options:
  - max_size:              maximum number of values stored in this cache (default: 1000)
  - max_key_size:          maximum key length (default: 250)
  - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
  - max_value:             maximum value length (default: 1048576)
  - pool_size:             pool size (default: 5)
  - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
  - retries:               number of retries (default: 3)
  - timeout:               default caching timeout in milliseconds (default: 1 minute)
    - failures:              number of failures before stop retrying (default: 5)
    - retry:                 retry timeout in milliseconds (default: 30 sec)
    - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)

References:

- *:discovery:*:*:1.0        (optional) IDiscovery services to resolve connection

Example:
	ctx := context.Background()

    lock := NewMemcachedLock[string]();
    lock.Configure(ctx, cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 11211,
    ));

    err := lock.Open(ctx, "123")
    if err != nil {
      ...
    }

    result, err := lock.TryAcquireLock(ctx, "123", "key1", 3000)
    if result {
    	// Processing...
    }
    err = lock.ReleaseLock(ctx, "123", "key1")
    // Continue...
*/
type MemcachedLock struct {
	*clock.Lock
	connectionResolver *ccon.ConnectionResolver
	// maxKeySize         int
	// maxExpiration      int64
	// maxValue           int64
	// poolSize           int
	// reconnect          int
	timeout int
	// retries            int
	// failures           int
	// retry              int
	remove bool
	//idle   int
	client *memcache.Client
}

// NewMemcachedLock method are creates a new instance of this lock.
func NewMemcachedLock() *MemcachedLock {
	c := &MemcachedLock{
		connectionResolver: ccon.NewEmptyConnectionResolver(),
		// maxKeySize:         250,
		// maxExpiration:      2592000,
		// maxValue:           1048576,
		// poolSize:           5,
		// reconnect:          10000,
		timeout: 5000,
		// retries:            5,
		// failures:           5,
		// retry:              30000,
		remove: false,
		//idle:   5000,
		client: nil,
	}
	c.Lock = clock.InheritLock(c)
	return c
}

// Configure method are configures component by passing configuration parameters.
//   - ctx context.Context
//   - config    configuration parameters to be set.
func (c *MemcachedLock) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.Lock.Configure(ctx, config)

	c.connectionResolver.Configure(ctx, config)

	// c.maxKeySize = config.GetAsIntegerWithDefault("options.max_key_size", c.maxKeySize)
	// c.maxExpiration = config.GetAsLongWithDefault("options.max_expiration", c.maxExpiration)
	// c.maxValue = config.GetAsLongWithDefault("options.max_value", c.maxValue)
	// c.poolSize = config.GetAsIntegerWithDefault("options.pool_size", c.poolSize)
	// c.reconnect = config.GetAsIntegerWithDefault("options.reconnect", c.reconnect)
	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	// c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	// c.failures = config.GetAsIntegerWithDefault("options.failures", c.failures)
	// c.retry = config.GetAsIntegerWithDefault("options.retry", c.retry)
	// c.remove = config.GetAsBooleanWithDefault("options.remove", c.remove)
	//c.idle = config.GetAsIntegerWithDefault("options.idle", c.idle)
}

// SetReferences method are sets references to dependent components.
//   - ctx context.Context
//   - references 	references to locate the component dependencies.
func (c *MemcachedLock) SetReferences(ctx context.Context, references cref.IReferences) {
	c.connectionResolver.SetReferences(ctx, references)
}

// IsOpen method are checks if the component is opened.
// Returns: true if the component has been opened and false otherwise.
func (c *MemcachedLock) IsOpen() bool {
	return c.client != nil
}

/// Open method are opens the component.
// Parameters:
//   - ctx context.Context
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *MemcachedLock) Open(ctx context.Context, correlationId string) error {
	connections, err := c.connectionResolver.ResolveAll(correlationId)

	if err == nil && len(connections) == 0 {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
	}

	if err != nil {
		return err
	}

	var servers []string = make([]string, 0)
	for _, connection := range connections {
		host := connection.Host()
		port := connection.Port()
		if port == 0 {
			port = 11211
		}

		servers = append(servers, host+":"+strconv.FormatInt(int64(port), 10))
	}

	// options = {
	//     maxKeySize: c.maxKeySize,
	//     maxExpiration: c.maxExpiration,
	//     maxValue: c.maxValue,
	//     poolSize: c.poolSize,
	//     reconnect: c.reconnect,
	//     timeout: c.timeout,
	//     retries: c.retries,
	//     failures: c.failures,
	//     retry: c.retry,
	//     remove: c.remove,
	//     idle: c.idle
	// };

	c.client = memcache.New(servers...)
	c.client.Timeout = time.Duration(c.timeout) * time.Millisecond
	//c.client.MaxIdleConns = c.idle

	return nil
}

// Close method are closes component and frees used resources.
// Parameters:
//   - ctx context.Context
//   - correlationId 	(optional) transaction id to trace execution through call chain.
//   - callback 			callback function that receives error or nil no errors occured.
func (c *MemcachedLock) Close(ctx context.Context, correlationId string) error {
	c.client = nil
	return nil
}

func (c *MemcachedLock) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// TryAcquireLock method are makes a single attempt to acquire a lock by its key.
// It returns immediately a positive or negative result.
// Parameters:
//    - ctx context.Context
//    - correlationId     (optional) transaction id to trace execution through call chain.
//    - key               a unique lock key to acquire.
//    - ttl               a lock timeout (time to live) in milliseconds.
//  Returns: a lock result or error.
func (c *MemcachedLock) TryAcquireLock(ctx context.Context, correlationId string, key string, ttl int64) (result bool, err error) {

	state, err := c.checkOpened(correlationId)
	if !state {
		return false, err
	}

	lifetimeInSec := ttl / 1000
	item := memcache.Item{
		Key:        key,
		Value:      []byte("lock"),
		Expiration: int32(lifetimeInSec),
	}
	err = c.client.Add(&item)

	if err != nil && err == memcache.ErrNotStored {
		return false, nil
	}
	return err == nil, err

}

// ReleaseLock method are releases prevously acquired lock by its key.
//    - ctx context.Context
//    - correlationId     (optional) transaction id to trace execution through call chain.
//    - key               a unique lock key to release.
//  Returns error or nil for success.
func (c *MemcachedLock) ReleaseLock(ctx context.Context, correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)
	if !state {
		return err
	}
	err = c.client.Delete(key)
	if err != nil && err == memcache.ErrCacheMiss {
		err = nil
	}
	return err
}
